src/Servers/Kestrel/Transport.Sockets/src/Internal/IOQueue.cs

```diff
@@ -43,14 +45,25 @@ void IThreadPoolWorkItem.Execute()
                     item.Callback(item.State);
                 }
 
-                lock (_workSync)
+                // All work done.
+                // Set _doingWork (0 == false) prior to checking IsEmpty to catch any missed work in interim.
+                _doingWork = 0;
+
+                // Check if there is work to do
+                if (_workItems.IsEmpty)
                 {
-                    if (_workItems.IsEmpty)
-                    {
-                        _doingWork = false;
-                        return;
-                    }
+                    // Nothing to do, exit.
+                    break;
                 }
+
+                // Is work, can we set it as active again (via atomic Interlocked), prior to scheduling?
```

Comment:
Well... only a single worker processing work; and more importantly a worker *always* to process the work, so it can't go idle if there is work. 

Can be more than one active, as another one might be leaving this loop.